jetty-server/src/main/java/org/eclipse/jetty/server/AsyncContentProducer.java

```diff
@@ -403,4 +404,53 @@ public String toString()
             _httpChannel
         );
     }
+
+    LockedSemaphore newLockedSemaphore()
+    {
+        return new LockedSemaphore();
+    }
+
+    /**
+     * A semaphore that assumes working under {@link AsyncContentProducer#lock()} scope.
+     */
+    class LockedSemaphore
```

Comment:
This looks like a useful class. Surprised it is not already available as a common extension of Condition.